File: java/ankit_kanyal/arrays/NextPermutations.java
// link- https://leetcode.com/problems/next-permutation/

class Solution {
    public void nextPermutation(int[] nums) {
        int n=nums.length;
    int cur=n-2;
if(n==1) return ;
while(cur>=0 && nums[cur]>=nums[cur+1]){
    cur--;
}

if(cur>=0){
        int j=nums.length-1;
        while(nums[j]<=nums[cur]) j--;
        swap(nums,cur,j);
    }
    reverse(nums,cur+1,nums.length-1);

return ;
}

static void swap(int []nums,int first,int second){
    int temp=nums[first];
    nums[first]=nums[second];
    nums[second]=temp;
}

static void reverse(int [] nums,int first,int second){
    while(first<second){
        swap(nums,first++,second--);
    }
}
}

File: javascript/ankit_kanyal/NextPermutations.js
// link- https://leetcode.com/problems/next-permutation/

var nextPermutation = function(nums) {
    let index=nums.length-2;
    while(index>=0 && nums[index]>=nums[index+1]) index--;
    if(index>=0){
        let j=nums.length-1;
        while(nums[j]<=nums[index]) j--;
        swap(nums,index,j);
    }
    reverse(nums,index+1,nums.length-1);
    
};

const swap=(arr,i,j)=>{
    let temp=arr[i];
    arr[i]=arr[j];
    arr[j]=temp;
}

const reverse=(arr,i,j)=>{
    while(i<j) swap(arr,i++,j--);
}
